### PR TITLE
[doc] Add ADR on model migration

### PR DESCRIPTION
**Pull request by mcharfadi**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+One important thing to consider is that BasicExtendedMetaData methods assume that the model is stored in XMl format, since we're using JSON format, we should be really cautious about using the default methods and might have to override most of them even if we don't use them all for migrating, a BasicJSonExtendedMetaData might even be needed.
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
+    private final String version;
+    private final List<IMigrationParticipant> iMigrationParticipants;
+    public MigrationExtendedMetaData(List<IMigrationParticipant> iMigrationParticipants, String version) {
+        this.iMigrationParticipants = Objects.requireNonNull(iMigrationParticipants);
+        this.version = version;
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: The version should be required too, we can initialize it to something arbitrary old for existing models lacking a version

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public EStructuralFeature getLocalElement(EClass eClass,  String namespace, String oldName) {
+        EStructuralFeature structuralFeature = eClass.getEStructuralFeature(oldName);
+        var migrationParticipants = iMigrationParticipants.stream().
+                filter(iMigrationParticipant -> iMigrationParticipant.canHandle(this.version))
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: Just use the version provided by the migration participant and check that its above the version currently consider. No need for a canHandle method here.

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+We will extend `ExtendedMetaData` with another class  `MigrationExtendedMetaData` that will be instantiated with the current version of the model expected by Sirius Web and a list of `IMigrationParticipant`.
+This class will also override the necessary methods from `BasicExtendedMetaData` that will allow us to migrate the model (like `getLocalElement`)
+
+One important thing to consider is that BasicExtendedMetaData methods assume that the model is stored in XMl format, since we're using JSON format, we should be really cautious about using the default methods and might have to override most of them even if we don't use them all for migrating, a BasicJSonExtendedMetaData might even be needed.
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
+    private final String version;
+    private final List<IMigrationParticipant> iMigrationParticipants;
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: `iMigrationParticipant(s)` -> `migrationParticipant(s)`

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+}
+----
+
+In each method overridden, we call the corresponding method from `IMigrationParticipant` if it has a lesser version than the one used to instantiate the `MigrationExtendedMetaData`.
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
+    boolean canHandle(String currentFileVersion);
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: No need for every single migration participant to compare their version and the current version, you can remove this `canHandle` method

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    String getMigrationParticipantVersion();
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: `getVersion` is enough, you know that it's a migration participant already

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    EStructuralFeature getLocalElement(EClass eClass,  String namespace, String oldName);
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: Use meaningful names contrary to the one provided by EMF. Here, it should be something like:

```
EStructuralFeature getStructuralFeature(EClass eClass, String featureName)
```

What is the `namespace`? Is that relevant for us?

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        return structuralFeature;
+    }
+}
+----
+
+In each method overridden, we call the corresponding method from `IMigrationParticipant` if it has a lesser version than the one used to instantiate the `MigrationExtendedMetaData`.
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: Shouldn't that be the signature of `IMigrationParticipant`?

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+[source,java]
+----
+@Service
+public class NameFeatureChangelMigrationParticipant implements IMigrationParticipant {
+
+    private static final String PARTICIPANT_VERSION = "202404221458";
+    @Override
+    public boolean canHandle(String currentFileVersion) {
+        return PARTICIPANT_VERSION.compareTo(currentFileVersion) < 0;
+    }
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: We don't need to code that in every migration participant

File: doc/adrs/142_add_support_for_emf_models_migration.adoc (hunk 6 above)
Comment (sbegaudeau) [reply doc/adrs/142_add_support_for_emf_models_migration.adoc]: You will need to ensure that we don't need to override every single methods of the interface `IMigrationParticipant`. If one needs to just rename an attribute, they should not be asked to retrieve classes, packages, etc.

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+----
+
+[source,java]
+----
+public class MigrationExtendedMetaData extends BasicExtendedMetaData {
+    boolean canHandle(String currentFileVersion);
+    String getMigrationParticipantVersion();
+    EStructuralFeature getLocalElement(EClass eClass,  String namespace, String oldName);
+}
+----
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: Duplicate? Or did I miss anything

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        for (IMigrationParticipant migrationParticipant : migrationParticipants) {
+            structuralFeature = migrationParticipant.getLocalElement(eClass, namespace, oldName);
+        }
+
+        return structuralFeature;
+    }
+}
+----
+
+In each method overridden, we call the corresponding method from `IMigrationParticipant` if it has a lesser version than the one used to instantiate the `MigrationExtendedMetaData`.
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: You are not talking about where this version will be stored? Something like a `migration` property at the root of the JSON document would be helpful but you need to serialize it and retrieve it which will require an update to sirius-emf-json. Preferably just a simple API to add whatever we need in the JSON document as a post process for exemple (but that will require a simple ADR on the EMFJson side). Anyway, you need to store something, somewhere for view, domain and other models that indicate the last version used for the migration. We could even store the name of the migration participant used to know quickly what has been the latest migration

```
{
  "json": {
    "version": "1.0",
    "encoding": "UTF-8"
  },
  "migration": {
    "lastMigrationPerformed": "NameFeatureChangelMigrationParticipant",
    "version": "202404221458"
  },
  "ns": {},
  "content": []
}
```

File: doc/adrs/142_add_support_for_emf_models_migration.adoc [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public EStructuralFeature getLocalElement(EClass eClass,  String namespace, String oldName) {
+        EStructuralFeature feature = null;
+        if (DomainPackage.eINSTANCE.getDomain().equals(eClass)) {
+            if (oldName.equals("name2")) {
+                feature = DomainPackage.eINSTANCE.getNamedElement_Name();
+            }
+        }
```
Comment (sbegaudeau) [inline doc/adrs/142_add_support_for_emf_models_migration.adoc]: Start thinking about the first concrete use case that this API should fulfill.

An attribute has disappeared (Node#labelExpression) and for those who had some expressions in this attribute, we need instead to create a new child (Node#insideLabel: InsideLabel) and set the expression on the newly created child (InsideLabel#labelExpression). I have a couple of other use cases in mind (there are a couple of attributes to rename for example but that's the first use case for the iteration).

No need to update the ADR for that use case, but when you will start implementing it, this is the first use case.

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: doc/adrs/142_add_support_for_emf_models_migration.adoc (hunk 7 above)
Comment (mcharfadi) [reply doc/adrs/142_add_support_for_emf_models_migration.adoc]: Yes indeed

File: doc/adrs/142_add_support_for_emf_models_migration.adoc (hunk 6 above)
Comment (mcharfadi) [reply doc/adrs/142_add_support_for_emf_models_migration.adoc]: Okay I thought I would be using the same name as the one provided by EMF. 
I also wanted to have the same methods in MigrationExtendedMetaData and the MigrationParticipant but I should only offer a few methods in MigrationParticipant to be implemented then, right ?

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**